bot/bot.py

```diff
@@ -39,6 +44,68 @@ def __init__(self, redis_session: RedisSession, **kwargs):
         self.loop.create_task(self.check_channels())
         self.loop.create_task(self.send_log(self.name, "Connected!"))
 
+        # Mapping for the main games leaderboard
+        # The key holds the name of the discord Cog and the value is the redis cache object
+        # for that specific game leaderboard
+        self.games_leaderboard: dict[str, tuple[RedisCache, ...]] = {}
+        self.auto_per_day_leaderboard_clear.start()
+
+    @loop(time=datetime.time.min)
+    async def auto_per_day_leaderboard_clear(self) -> None:
+        """Loop for clearing the per day leaderboard redis cache at UTC midnight."""
+        for _, per_day_lb in self.games_leaderboard.values():
+            await per_day_lb.clear()
+
+    def ensure_leaderboard(self, cog: Cog) -> None:
+        """Ensure that per day leaderboard and total leaderboard redis cache are available for the cog."""
+        if self.games_leaderboard.get(cog.qualified_name):
+            return
+
+        # This cache contains the leaderboard for game `cog`
+        # RedisCache[int: int]
+        # Where the key is the discord user ID, and the values if the total
+        # points registered for that user.
+        _leaderboard_cache = RedisCache(namespace=f"{cog.qualified_name}-total")
+
+        # This cache contains the per day leaderboard, same specifies
+        _per_day_leaderboard = RedisCache(namespace=cog.qualified_name)
+
+        self.games_leaderboard[cog.qualified_name] = (_leaderboard_cache, _per_day_leaderboard)
+
+    async def increment_points(self, cog: Cog, user: discord.Member, points: int = DEFAULT_POINTS) -> None:
+        """
+        Increment points for a user in the leaderboard for game `cog` by `points`.
+
+        If max points for today are hit for that particular game cog, then set today's points to the max
+        points and add the difference of today's points and maximum daily possible points to the overall
+        points for the user.
+        """
+        try:
+            _leaderboard_cache, _per_day_leaderboard = self.games_leaderboard[cog.qualified_name]
+        except KeyError:
+            log.error(
+                f"No leaderboards found for cog {cog.qualified_name}. Make sure to run `ensure_leaderboard` "
+                f"before using this utility function."
+            )
+            return
```

Comment:
I want to keep the two separate, like we run `self.bot.ensure_leaderboard(ext)` in the ext `__init__` like how we need to run [`ctx.ensure_objects()`](https://click.palletsprojects.com/en/8.0.x/complex/#ensuring-object-creation) before storing values in `ctx.obj[]` in [pallets/click](https://github.com/pallets/click).

Secondly, this would unnecessarily be running every time your cog calls `increment_points` whereas keeping this in the `__init__` would only call it during the cog setup. 

You can see the `coinflip` example, I included in this PR.